---
title: "No models, just interpretable rules"
author: "callitmagic"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
    code_folding: show
    code_download: true
---

### Software used

[The MEME Suite](https://meme-suite.org/meme/)

```{bash}
/home/ubuntu/bin/meme -version
```

### Setup

```{r setup, message = FALSE, error = FALSE}
knitr::opts_chunk$set(message = FALSE, error = FALSE)
sapply(list.files("functions/", full.names = TRUE), source)
```

#### Session info and constants

```{r}
require("Biostrings")
require("magrittr")
require("dplyr")
require("readr")
require("purrr")
require("parallel")
require("glue")
require("XML")
require("tidyr")
set.seed(53)
GENOME_PATH <- "data/genome/"
TRAIN_PATH <- "data/train_unified/"
DATA_PATHS <- list(
  "lb" = list(
    train = "data/raw_data/IBIS.train_data.Leaderboard.v2/",
    test = "data/raw_data/IBIS.test_data.Leaderboard.v3/",
    subm = "data/raw_data/Leaderboard_aaa_submissions.v3/",
    is_genome = c("GABPA", "PRDM5", "SP140", "ZNF362", "ZNF407"),
    is_artificial = c("LEF1", "NACC2", "RORB", "TIGD3", "NFKB1")
  ),
  "fin" = list(
    train = "data/raw_data/IBIS.train_data.Final.v1/train/",
    test = "data/raw_data/IBIS.test_data.Final.v1/",
    subm = "data/raw_data/Final_aaa_submissions.v1/",
    is_genome = c("CAMTA1", "LEUTX", "MYF6", "PRDM13", "SALL3", "USF3", "ZBED2", "ZBED5", "ZNF20", "ZNF251", "ZNF367", "ZNF395", "ZNF493", "ZNF518B", "ZNF648"),
    is_artificial = c("GCM1", "MKX", "MSANTD1", "MYPOP", "SP140L", "TPRX1", "ZFTA", "CREB3L3", "FIZ1", "ZBTB47", "ZNF286B", "ZNF500", "ZNF721", "ZNF780B", "ZNF831")
  )
)
DATA_TYPES <- names(DATA_PATHS)
sessionInfo()
sapply(list.files("functions/", full.names = TRUE), source)
```

#### Genome

```{bash, class.source="bg-warning", eval=FALSE}
cd data/genome/
for i in {1..22}; do file="chr${i}.fa.gz"; [ ! -f "$file" ] && wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/$file; done
cd ../../
```

#### Data download

```{bash, class.source="bg-warning", results = "hide", eval = FALSE}
cd data/raw_data/
wget https://ibis.autosome.org/IBIS_data/IBIS.test_data.Leaderboard.v3.zip
wget https://ibis.autosome.org/IBIS_data/IBIS.train_data.Leaderboard.v2.zip
wget https://ibis.autosome.org/IBIS_data/IBIS.train_data.Final.v1.zip
wget https://ibis.autosome.org/IBIS_data/IBIS.test_data.Final.v1.zip
wget https://ibis.autosome.org/IBIS_data/Leaderboard_aaa_submissions.v3.zip
wget https://ibis.autosome.org/IBIS_data/Final_aaa_submissions.v1.zip

for f in *.zip; do dir="${f%.zip}"; mkdir "$dir" && unzip "$f" -d "$dir" && rm "$f"; done
cd ../../
```

#### Data conversion

```{r}
rewrite <- FALSE
data <- tibble()
for (pt in DATA_TYPES) {
  path <- glue::glue("{TRAIN_PATH}CHS_GHTS_{pt}.RDS")
  if (file.exists(path) & !rewrite) {
    print(paste(path, "exists"))
    next
  }
  saveRDS(rbind(
    merge_peak_files(glue::glue("{DATA_PATHS[[pt]]$train}/CHS"), GENOME_PATH),
    merge_peak_files(glue::glue("{DATA_PATHS[[pt]]$train}/GHTS"), GENOME_PATH)
  ), path)
}

for (pt in DATA_TYPES) {
  path <- glue::glue("{TRAIN_PATH}PBM_{pt}.RDS")
  if (file.exists(path) & !rewrite) {
    print(paste(path, "exists"))
    next
  }
  saveRDS(merge_peak_files(glue::glue("{DATA_PATHS[[pt]]$train}/PBM"), GENOME_PATH), path)
}

for (pt in DATA_TYPES) {
  path <- glue::glue("data/train_unified/HTS_{pt}.RDS")
  if (file.exists(path) & !rewrite) {
    print(paste(path, "exists"))
    next
  }
  data <- merge_seqs(glue::glue("{DATA_PATHS[[pt]]$train}/HTS"))
  rownames(data) <- NULL
  saveRDS(data, glue::glue("data/train_unified/HTS_{pt}.RDS"))
}

for (pt in DATA_TYPES) {
  path <- glue::glue("data/train_unified/SMS_{pt}.RDS")
  if (file.exists(path) & !rewrite) {
    print(paste(path, "exists"))
    next
  }
  data <- merge_seqs(glue::glue("{DATA_PATHS[[pt]]$train}/SMS"))
  rownames(data) <- NULL
  saveRDS(data, path)
}
rm(data)
gc()
```

## PWM track

### Generating motifs - parameters grids

```{r}
param_grid <- expand.grid(
  length = c(5, 10, 15, 30, NA),
  GHTS = c(TRUE, FALSE),
  CHS = c(TRUE, FALSE),
  meme_rev = c(TRUE, FALSE),
  pileup = c(25, 50, 75, 90, NA),
  maxw = c(15, 30),
  seed = 53
)
param_grid %<>% filter(GHTS | CHS)
param_grid %<>% filter(length*2 >= maxw | is.na(length))
additional_grid <- expand.grid(
  length = c(5),
  GHTS = c(TRUE, FALSE),
  CHS = c(TRUE, FALSE),
  meme_rev = c(FALSE),
  pileup = c(25, NA),
  maxw = c(30),
  seed = 53
)
additional_grid %<>% filter(GHTS | CHS)
param_grid %<>% rbind(additional_grid) 
param_grid$n <- 1:nrow(param_grid)
saveRDS(param_grid, "data/params/genome_param_grid.RDS")
param_grid <- expand.grid(
  SMS = c(TRUE, FALSE),
  PBM = c(TRUE, FALSE),
  HTS = c(TRUE, FALSE),
  PBM_filter = c("max", "2sd", NA),
  PBM_subset = c("QNZS", "SD", NA),
  HTS_subset = c("C4", "C3", NA),
  meme_rev = c(TRUE, FALSE),
  seed = 53,
  stringsAsFactors = FALSE
)
param_grid %<>% filter(SMS | PBM | HTS)
param_grid %<>% filter(!(SMS & PBM & HTS & !is.na(PBM_subset)))
param_grid %<>% filter(!(SMS & PBM & HTS & !is.na(PBM_filter)))
param_grid %<>% filter((!PBM & (is.na(PBM_filter) & is.na(PBM_subset))) | (PBM & !is.na(PBM_filter)))
param_grid %<>% filter((!HTS & is.na(HTS_subset)) | HTS)
param_grid %<>% rbind(expand.grid(
  SMS = c(FALSE),
  PBM = c(TRUE),
  HTS = c(FALSE),
  PBM_filter = c(NA),
  PBM_subset = c("QNZS", "SD", NA),
  HTS_subset = c(NA),
  meme_rev = c(TRUE, FALSE),
  seed = 53,
  stringsAsFactors = FALSE
))

additional_grid <- expand.grid(
  SMS = c(FALSE),
  PBM = c(TRUE),
  HTS = c(TRUE, FALSE),
  PBM_filter = c("5", "10"),
  PBM_subset = c(NA),
  HTS_subset = c("C3", NA),
  meme_rev = c(TRUE),
  seed = 53,
  stringsAsFactors = FALSE
)
param_grid %<>% rbind(additional_grid)
param_grid$n <- 1:nrow(param_grid)
param_grid %<>% filter(!SMS)
saveRDS(param_grid, "data/params/artificial_param_grid.RDS")
```

### Generating motifs - genome data

```{r, results = "hide", eval = FALSE}
param_grid <- readRDS("data/params/genome_param_grid.RDS")
row_list <- split(param_grid, seq(nrow(param_grid)))
for (type in DATA_TYPES) {
  data_ <- get_train(exp = "CHS_GHTS", pt = type)
  results <- lapply(row_list, function(row) process_row.genome(row, data_, pt = type))
  saveRDS(results, glue::glue("timings/{type}_genome_param_grid.RDS"))
  files <- list.files(glue::glue("meme_comparison/data/tmp_memedir_{type}"), recursive = TRUE, full.names = TRUE)
  file.remove(files[!grepl("meme\\.xml|meme\\.txt$", files)])
}
```

### Generating motifs - artificial data

```{r, results = "hide", eval = FALSE}
param_grid <- readRDS("data/params/artificial_param_grid.RDS")
row_list <- split(param_grid, seq(nrow(param_grid)))
for (type in DATA_TYPES) {
  data_ <- list(SMS = get_train(exp = "SMS", pt = type), PBM = get_train(exp = "PBM", pt = type), HTS = get_train(exp = "HTS", pt = type))
  results <- lapply(row_list, function(row) process_row.artificial(row, data_, pt = type))
  saveRDS(results, glue::glue("timings/{type}_artificial_param_grid.RDS"))
  files <- list.files(glue::glue("meme_comparison/data/tmp_memedir_{type}"), recursive = TRUE, full.names = TRUE)
  file.remove(files[!grepl("meme\\.xml|meme\\.txt$", files)])
}
```

### Cleaning motifs

```{r}
evalue_threshold <- 0.05
filter_meme_file <- function(file_path, evalue_threshold) {
  xml_doc <- xmlParse(file_path)
  motifs <- getNodeSet(xml_doc, "//motif")
  for (motif in motifs) {
    evalue <- as.numeric(xmlGetAttr(motif, "e_value"))
    if (evalue > evalue_threshold) {
      removeNodes(motif)
    }
  }
  if (length(getNodeSet(xml_doc, "//motif")) > 0) {
    saveXML(xml_doc, file = sub("meme.xml$", "filtered_meme.xml", file_path))
  }
  file.remove(file_path)
}
for (type in DATA_TYPES) {
  meme_files <-
    list.files(
      path = glue::glue("meme_comparison/data/tmp_memedir_{type}/"),
      pattern = "^meme.xml$",
      recursive = TRUE,
      full.names = TRUE
    )
  if (length(meme_files) > 0) {
    res <- lapply(meme_files, filter_meme_file, evalue_threshold = evalue_threshold)
  }
  cat(glue::glue("Filtered {type} meme.xml files based on the e-value threshold."), "\n")
}
```

### AME comparison

#### Control sequences

```{r, eval = FALSE}
dir_ <- "meme_comparison/references/"
ind <- 20

for (type in DATA_TYPES) {
  references <- rbind(
    get_train(exp = "CHS_GHTS", pt = type) %>% select(tf, seq, method),
    get_train(exp = "PBM", pt = type) %>%
      mutate(method = gsub("_.*", "", f_name)) %>%
      group_by(method) %>%
      mutate(
        mean_sig = mean(mean_signal_intensity),
        sd_sig = sd(mean_signal_intensity),
        th = mean_sig + 2 * sd_sig
      ) %>%
      ungroup %>%
      filter(mean_signal_intensity > th) %>%
      select(tf, seq = pbm_sequence) %>% mutate(method = "PBM"),
    get_train(exp = "HTS", pt = type) %>%
      filter(grepl("R0_C4", f_name) | grepl("R1_C4", f_name)) %>%
      select(tf, seq) %>%
      mutate(method = "HTS") %>% group_by(tf) %>% sample_frac(0.5) %>% ungroup
  )
  references %<>% filter(!grepl("N", seq))
  references$n <- 1:nrow(references)
  saveRDS(references, glue::glue("{dir_}full_{type}.RDS"))
  for (tf_ in unique(references$tf)) {
    pos <- glue::glue("{dir_}/{tf_}_positive_{type}.fasta")
    if (file.exists(pos)) next
    writeLines(paste0(
      paste0(">", references$n[references$tf == tf_], "\n"), references$seq[references$tf == tf_], 
      collapse = "\n"), 
      pos)
    neg <- glue::glue("{dir_}/{tf_}_negative_{type}.fasta")
    if (file.exists(neg)) next
    writeLines(paste0(
      paste0(">", references$n[references$tf != tf_], "\n"), references$seq[references$tf != tf_], 
      collapse = "\n"), 
      neg)
  }
  references <- references %>% mutate(seq = substr(seq, nchar(seq)/2 - ind, nchar(seq)/2 + ind))
  is_genome <- DATA_PATHS[[type]]$is_genome
  references$is_genome <- references$tf %in% is_genome
  for (tf_ in unique(references$tf)) {
    pos <- glue::glue("{dir_}/{tf_}_sep_cut_positive_{type}.fasta")
    if (file.exists(pos)) next
    writeLines(paste0(
      paste0(">", references$n[references$tf == tf_], "\n"), references$seq[references$tf == tf_], 
      collapse = "\n"), 
      pos)
    neg <- glue::glue("{dir_}/{tf_}_sep_cut_negative_{type}.fasta")
    if (file.exists(neg)) next
    writeLines(paste0(
      paste0(">", references$n[references$tf != tf_ & (references$is_genome == tf_ %in% is_genome)], "\n"), 
      references$seq[references$tf != tf_ & (references$is_genome == tf_ %in% is_genome)], 
      collapse = "\n"), 
      neg)
  }
  references <- rbind(
    get_train(exp = "CHS_GHTS", pt = type) %>% select(tf, seq, method),
    get_train(exp = "PBM", pt = type) %>%
      mutate(method = gsub("_.*", "", f_name)) %>%
      group_by(method) %>%
      mutate(mean_sig = mean(mean_signal_intensity), sd_sig = sd(mean_signal_intensity), th = mean_sig + 2*sd_sig) %>%
      ungroup %>%
      filter(mean_signal_intensity > th) %>%
      select(tf, seq = pbm_sequence) %>% mutate(method = "PBM"),
    get_train(exp = "HTS", pt = type) %>%
      filter(grepl("R0_C4", f_name) | grepl("R0_C3", f_name) | grepl("R1_C4", f_name) | grepl("R1_C3", f_name)) %>%
      select(tf, seq) %>%
      mutate(method = "HTS") %>% group_by(tf) %>% sample_frac(0.5) %>% ungroup
  )
  ind <- 20
  references %<>% filter(!grepl("N", seq))
  references$n <- 1:nrow(references)
  saveRDS(references, glue::glue("{dir_}full_ext_{type}.RDS"))
  references <- references %>% mutate(seq = substr(seq, nchar(seq)/2 - ind, nchar(seq)/2 + ind))
  references$is_genome <- references$tf %in% is_genome
  for (tf_ in unique(references$tf)) {
    pos <- glue::glue("{dir_}/{tf_}_sep_cut_ext_positive_{type}.fasta")
    if (file.exists(pos)) next
    writeLines(paste0(
      paste0(">", references$n[references$tf == tf_], "\n"), references$seq[references$tf == tf_], 
      collapse = "\n"), 
      pos)
    neg <- glue::glue("{dir_}/{tf_}_sep_cut_ext_negative_{type}.fasta")
    if (file.exists(neg)) next
    writeLines(paste0(
      paste0(">", references$n[references$tf != tf_ & (references$is_genome == tf_ %in% is_genome)], "\n"), 
      references$seq[references$tf != tf_ & (references$is_genome == tf_ %in% is_genome)], 
      collapse = "\n"), 
      neg)
  }
}
```

#### AME processing

```{r, results = "hide"}
dir_ <- "meme_comparison/references/"
names_g <- readRDS("data/params/genome_param_grid.RDS")
names_g <- unlist(lapply(split(names_g, seq(nrow(names_g))), function(row) paste(row, collapse = "_")))
names_a <- readRDS("data/params/artificial_param_grid.RDS")
names_a <- unlist(lapply(split(names_a, seq(nrow(names_a))), function(row) paste(row, collapse = "_")))

args <- commandArgs(trailingOnly=TRUE)
ind <- if (is.na(as.numeric(args[1]))) args[1] else as.numeric(args[1])

for (type in DATA_TYPES) {
  all_tfs <- c(DATA_PATHS[[type]]$is_genome, DATA_PATHS[[type]]$is_artificial)
  if (is.numeric(ind)) {
    tfs <- all_tfs[(ind*5 + 1):(ind*5 + 5)]
    tfs <- tfs[!is.na(tfs)]
  } else {
    tfs <- all_tfs
  }
  if (length(tfs) == 0) next
  comparison_grid <- rbind(
    expand.grid(tf = DATA_PATHS[[type]]$is_genome, 
                params = names_g, 
                stringsAsFactors = FALSE),
    expand.grid(tf = DATA_PATHS[[type]]$is_artificial, 
                params = names_a, 
                stringsAsFactors = FALSE)
  )
  comparison_grid %<>% filter(tf %in% tfs)
  comparison_grid$params <- file.path(glue::glue("meme_comparison/data/tmp_memedir_{type}/"), comparison_grid$params)
  for (tf_ in unique(comparison_grid$tf)) {
    print(tf_)
    suff_ <- "_sep_cut"
    if (tf_ %in% DATA_PATHS[[type]]$is_genome) suff_ <- ""
    print(suff_)
    comparison_grid_pt <- comparison_grid %>% filter(tf == tf_)
    comparison_grid_pt$bucket <- ceiling(seq_along(comparison_grid_pt$tf) / 20)
    row_list <- split(comparison_grid_pt, comparison_grid_pt$bucket)
    results <- lapply(row_list, function(row) process_ame(row, dir_, suff = suff_, type = type))
  }
}
```

#### Getting best PWMs

```{r}
for (type in DATA_TYPES) {
  is_genome <- DATA_PATHS[[type]]$is_genome
  path <- glue::glue("meme_comparison/data/ame_{type}/")
  data_ <- lapply(list.files(path, full.names = TRUE), readRDS)
  data_ <- data_[sapply(data_, length) == 6]
  if (length(data_) > 1) {
    data_ %<>% Reduce(rbind, .)
  }
  data_$tf <- basename(dirname(data_$X2))
  path <- glue::glue("meme_comparison/data/ame_sep_cut_{type}/")
  data_cut <- lapply(list.files(path, full.names = TRUE), readRDS) # was ame_sep
  data_cut <- data_cut[sapply(data_cut, length) == 6]
  if (length(data_cut) > 1) {
    data_cut %<>% Reduce(rbind, .)
  }
  data_cut$tf <- basename(dirname(data_cut$X2))
  # filtering <- readRDS(glue::glue("data/params/artificial_param_grid.RDS"))
  # filtering %<>% filter(!PBM_filter %in% c("5", "10"))
  # allowed_paths <- unlist(lapply(split(filtering, seq(nrow(filtering))), function(row) paste(row, collapse = "_")))
  # data_cut %<>% filter(sapply(X2, function(x) any(sapply(allowed_paths, function(path) grepl(pattern = path, x)))))
  data <- rbind(data_ %>% filter(tf %in% is_genome), data_cut %>% filter(!(tf %in% is_genome)))
  data %<>% group_by(tf) %>% mutate(score_ind = score >= quantile(score, 0.75)) %>% ungroup
  best_vars_ext <- data %>% group_by(tf) %>% arrange(desc(score)) %>% distinct(X3, .keep_all = TRUE) %>% 
    slice_head(n = 12) %>% 
    ungroup
  saveRDS(best_vars_ext, glue::glue("meme_comparison/preds/best_vars_ext_selected_{type}12.RDS"))
  
  st <- data %>% group_by(tf) %>% 
    arrange(desc(score)) %>% 
    distinct(X3, .keep_all = TRUE) %>% 
    slice_head(n = 4) %>% 
    mutate(n = row_number()) %>% ungroup
  collect_pwms_partial(st, glue::glue("meme_comparison/preds/{type}_PWM_{Sys.time()}_{type}.txt"))
}
```

## AAA track

#### Getting predictions from the best PWMs

##### Getting tags and tfs from submission templates

```{r, eval = FALSE}
for (type in DATA_TYPES) {
  experiments <- list.files(DATA_PATHS[[type]]$subm, pattern = "random", full.names = TRUE)
  for (experiment in experiments) {
    exp_ <- gsub("_aaa_random.tsv", "", gsub(paste0(DATA_PATHS[[type]]$subm, "/"), "", experiment))
    exp <- read_delim(experiment) 
    saveRDS(exp$tag, glue::glue("data/submissions/tags_{exp_}_{type}.RDS"))
    saveRDS(colnames(exp)[-1], glue::glue("data/submissions/tfs_{exp_}_{type}.RDS"))
    rm(exp)
  }
}
```

##### Predict with FIMO

```{r, results = "hide"}
base_command <- "/home/ubuntu/bin/fimo"
for (type in DATA_TYPES) {
  best_vars <- readRDS(glue::glue("meme_comparison/preds/best_vars_ext_selected_{type}12.RDS"))
  experiments <- list.files(DATA_PATHS[[type]]$subm, pattern = "random", full.names = TRUE)
  exp_name <- glue::glue("final_{type}")
  tryCatch({
    for (experiment in experiments) {
      exp_ <- gsub("_aaa_random.tsv", "", gsub(paste0(DATA_PATHS[[type]]$subm, "/"), "", experiment))
      tfs_ <- readRDS(glue::glue("data/submissions/tfs_{exp_}_{type}.RDS"))
      for (tf_ in tfs_) {
        fimo_dir <- glue::glue("tmp/fimo_out_{type}/{exp_name}/{tf_}/{exp_}")
        if (!dir.exists(fimo_dir)) dir.create(fimo_dir, recursive = TRUE)
        best_vars_sub <- best_vars[best_vars$tf == tf_, ]
        for (i in 1:nrow(best_vars_sub)) {
          fimo_out_dir <- paste0("meme_comparison/data/fimo_out_", type, "/", exp_name, "/", exp_, "/")
          fimo_out_fname <- paste0(fimo_out_dir, basename(dirname(dirname(best_vars_sub$X2[i]))), "_", best_vars_sub$X4[i], "_", tf_, ".RDS")
          if (file.exists(fimo_out_fname)) next
          run_fimo <- glue::glue(
              "{base_command} --motif {best_vars_sub$X3[i]} --max-stored-scores 10000000 --oc {fimo_dir} {best_vars_sub$X2[i]} {DATA_PATHS[[type]]$test}/{exp_}_participants.fasta")
          system(run_fimo)
          for (file in list.files(fimo_dir, full.names = TRUE)[!grepl("tsv$", list.files(fimo_dir))]) {
            file.remove(file)
          }
          f_name <- glue::glue("{fimo_dir}/fimo.tsv")
          if (!dir.exists(fimo_out_dir)) dir.create(fimo_out_dir, recursive = TRUE)
          if (file.exists(f_name)) {
            dt <- read_delim(f_name)
            dt %<>% filter(!is.na(sequence_name))
            dt$prob_1 <- exp(dt$score) / (1 + exp(dt$score))
            dt$prob_2 <- 1 - dt$`q-value`
            dt %<>% filter(`p-value` < 0.001)
            dt$tf <- tf_
            if (nrow(dt) > 0) {
              saveRDS(dt, fimo_out_fname)
            }
          }
        }
      }
    }
  },
  error = function(cond)
    print(cond))
}
```

##### Collect FIMO predictions

```{r, results = "hide"}
for (type in "lb") {
  experiments <- list.files(DATA_PATHS[[type]]$subm, pattern = "random", full.names = TRUE)
  exp_name <- glue::glue("final_{type}_raw")
  best_vars <- readRDS(glue::glue("meme_comparison/preds/best_vars_ext_selected_{type}12.RDS"))
  for (experiment in experiments) {
    exp_ <- gsub("_aaa_random.tsv", "", gsub(paste0(DATA_PATHS[[type]]$subm, "/"), "", experiment))
    files <- list.files(glue::glue('meme_comparison/data/fimo_out_{type}/{if(grepl("_raw", exp_name)) gsub("_raw", "", exp_name) else exp_name}/{exp_}/'), full.names = TRUE)
    tfs <- readRDS(glue::glue("data/submissions/tfs_{exp_}_{type}.RDS"))
    exp <- tibble(tag = readRDS(glue::glue("data/submissions/tags_{exp_}_{type}.RDS")))
    for (tf_ in tfs) {
      fldr <- glue::glue("meme_comparison/data/partial_preds/{type}")
      if (!dir.exists(fldr)) dir.create(fldr, recursive = TRUE)
      p_files <- files[grepl(paste0("_", tf_, ".RDS"), files)]
      best_vars_pt <- best_vars %>% filter(tf == tf_)
      allowed_paths <- paste0(basename(dirname(dirname(best_vars_pt$X2))), "_", best_vars_pt$X4, "_", best_vars_pt$tf)
      p_files <- p_files[sapply(p_files, function(x) any(sapply(allowed_paths, grepl, x)))]
      f_name <- glue::glue('{fldr}/{exp_}_{tf_}_{exp_name}_preds.RDS')
      if (file.exists(f_name)) next
      data <- lapply(p_files, function(x) {
        dt <- readRDS(x)
        dt$tf <- gsub(".RDS", "", gsub(".*_", "", x))
        dt
      })
      data %<>% Reduce(rbind, .)
      data %<>% group_by(sequence_name, tf) %>% summarize(
        # mean_score = median(score),
        max_score = max(score)#,
        # mean_prob_1 = mean(prob_1),
        # max_prob_1 = max(prob_1),
        # mean_prob_2 = mean(prob_2),
        # max_prob_2 = max(prob_2),
        # q_max = max(`q-value`, na.rm = TRUE),
        # q_min = min(`q-value`, na.rm = TRUE),
        # q_med = median(`q-value`, na.rm = TRUE)
      )
      names(data)[1] <- "tag"
      data <- exp %>% left_join(data %>% group_by(tag, tf) %>% summarize(score = max(max_score), .groups = "drop") %>% spread(tf, score))
      data[is.na(data)] <- 0
      if(!grepl("raw", exp_name)) data %<>% mutate(across(where(is.numeric), ~(. - min(.)) / (max(.) - min(.) + 1)))
      saveRDS(data, f_name)
      rm(data)
    }
    fldr <- glue::glue("meme_comparison/preds/{type}/")
    f_name <- glue::glue("{fldr}/{exp_}_{exp_name}.RDS")
    if (file.exists(f_name)) next
    files <- list.files(glue::glue("meme_comparison/data/partial_preds/{type}"), pattern = glue::glue("^{exp_}"), full.names = TRUE)
    files <- files[grepl(paste0(exp_name, "_preds.RDS"), files)]
    all <- Reduce(cbind, lapply(files, function(file) {dt <- readRDS(file); dt[, 2]}))
    if (grepl("raw", exp_name)) {
      min_val <- min(all)
      max_val <- max(all)
      all[all == 0] <- min_val - 1
      min_val <- min(all)
      all <- (all - min_val) / (max_val - min_val)
    }
    all <- cbind(exp, all)
    if (!dir.exists(fldr)) dir.create(fldr, recursive = TRUE)
    # saveRDS(all, f_name)
    all %<>% mutate(across(where(is.numeric), ~sprintf("%.5f", round(., 5))))
    write_delim(all, gsub(".RDS", ".tsv", f_name), delim = "\t")
    system(glue::glue('gzip -f {gsub(".RDS", ".tsv", f_name)}'))
  }
}
```
